go/vt/sqlparser/sql.y

```diff
@@ -1288,39 +1778,70 @@ match_option:
     $$ = QueryExpansionStr
  }
 
+charset:
+  ID
+{
+    $$ = string($1)
+}
+| STRING
+{
+    $$ = string($1)
+}
 
 convert_type:
-  charset
+  BINARY length_opt
```

Comment:
Yeah... I made this note in the very first commit on this PR (thanks for reminding me).

> This change required modifying our handling of CONVERT statements since there are now several new keywords and we actually have grammar rules for the data types. As such the new grammar is more restrictive than it was before, although it passes all tests and should cover all the cases listed in the mysql 5.7 reference docs.